### DISCUSSION TITLE
VIM key bindings in editor

### DISCUSSION THREAD

**Discussion by cs2018ncsa** [tconbeer/harlequin#371] [Ideas (Feature Requests)]
Is there a way to set the editor to use VIM config?

**Comment by tconbeer** [comment]
Not currently. I'd like to add an option in the future. Can you (or someone) please compile a list of modes and bindings? I'm not a vim user.

**Reply by tconbeer** [reply]
One wrinkle; Textual plans to make key bindings configurable in the future, but that isn't supported today. There is a [workaround](https://github.com/Textualize/textual/discussions/2006) that we could consider, but I'm inclined to wait for the upstream features to get fleshed out more. But PRs are welcome!

**Comment by avilesj** [comment]
@tconbeer **Thank you** for building this. I will be watching closely for vim mode!

**Comment by joshtemple** [comment]
@tconbeer this [is the official list](https://vimhelp.org/index.txt.html), but I'm sure there's a subset of commands that would work and would be a good start.

**Comment by tconbeer** [comment]
Big update in this realm: configurable key bindings just shipped with v1.22.0. See the docs: https://harlequin.sh/docs/keymaps/index

However, supporting Vim motions requires a LOT more work, since Textual bindings don't natively support multiple key presses (and a host of other challenges). If anyone has Vim-ish bindings that they are able to configure post-v1.22, please share your config here (or distribute those as a plug-in). Thanks!

**Comment by joshtemple** [comment]
@tconbeer, I'm not at all a Vim power user, but as I see it, implementing at a minimum Normal and Insert modes will be a requirement for users to configure their own Vim-ish bindings.

Currently it seems not possible to implement Vim bindings because many commands are single keypresses in Normal mode, and in Harlequin you'll want to treat those single keypresses as editing the active buffer (like `j`, `k`, etc.).

If you could implement a toggle for those two modes, the community could probably do more here with building the keymap as a plugin.

Here's how a basic set of commands could work. This wouldn't require adding any new commands, just mapping keys differently while in Normal mode.

### In Normal Mode:

 * **`i`: Enter Vim Insert Mode**
 * `j`: Cursor Up
 * `k`: Cursor Down
 * `h`: Cursor Left
 * `l`: Cursor Right
 * `b`: Cursor Word Left
 * `w`: Cursor Word Right
 * `0`: Cursor Line Start
 * `SHIFT + 4 [$]`: Cursor Line End
 * `g`: Cursor Doc Start
 * `SHIFT + g [G]`: Cursor Doc End

### In Insert Mode:

 * **`ESC`: Enter Vim Normal Mode**
 * Maintain current editor behavior

You're right that you'll still need multi-keypress support to support motions like delete word (`dw`) or change word (`cw`), but I think you can solve for a lot of the navigation with single key presses.

**Comment by miscbits** [comment]
The major modes that I want to see implemented are normal mode where basic navigation happens, visual mode which allows you to select blocks of text, and insert mode which allows you to actually type information into the editor. There is also replace mode which is similar to insert mode but it is for replacing content.

Less important, but would still be command mode which allows you to type in custom commands and functions. This can allow you to run commands that you don't have keybinds for (like opening a file or directory, connecting to a database, refreshing a folder/bucket)

If at the very least, navigational/normal mode could be implemented it would be a huge improvement in terms of usability.

**Comment by cs2018ncsa** [comment]
@tconbeer looks like textualize has the keymap functionality you were waiting for.

**Comment by ghost** [comment]
@tconbeer , got myself a programmable keyboard and trying **hard** to get keybindings set in the spirit of Vim.  And feeling only partially successful; dealing with what feels like bugs, but maybe I'm missing something.

For example: 

- Navigation: CTL + h,j,k,l
- Visual: CTL + H,J,K,L
- Run Query: CTL + w
- Next Buffer: Tab
- Prev Buffer: Shift Tab

Some of these overrides work, and others don't. I've ensured there aren't any duplicates in both my `harlequin --keys` and my .toml.

I could provide specifics if needed, but I'm having to result to default mappings (vscode) and keyboard layer programming - which works but is a hassle and requires holding layer-mode keys in Normal/Visual.  Any thoughts/guidance?

**Comment by tothalex** [comment]
@tconbeer is there any update for this feature?

**Comment by xheisenbugx** [comment]
any updates on this feature?

**Comment by Mac-lp3** [comment]
Literally the only thing stopping from maining Harley. You could start pushing hot garbage after this and I would still deal with it and defend this project.

The bare minimum I need is:

1. introduce the concept of "modes" (insert mode vs normal aka "navigation" mode), and
2. mappings for "focus on the window to the left/up/down/right," in addition to the current keybindings for "focus on the Query Editor/Results Viewer/etc."

In normal mode, you have key mappings to move the cursor between windows, and key mappings to move the cursor around the text inside of a window. 

For the former, I would just copy over my VIM key maps to navigate vertical/horizontal splits.

For the latter, support for the billions of ways to move the cursor would be nice, but I'd cope with smashing `w` and `b` until I put a hole in my keeb.

**Reply by ghost** [reply]
Seriously, same. I stopped using Harequin altogether for this reason and have been on dadbod for a while now. Tried Harlequin in a nvim terminal, too clunky.  Sometimes I’ll use the vim extension in DBeaver or VS Code, but I’d much rather stay in the terminal.

**Reply by joanofxyz** [reply]
as a heavy vim user, I honestly think the lazygit approach to vim style keybindings and layout is great for TUIs (no <C-W>[motion] necessary)

**Reply by devPeralta** [reply]
> as a heavy vim user, I honestly think the lazygit approach to vim style keybindings and layout is great for TUIs (no [motion] necessary)

lazygit dont write code, harlequin does

**Comment by mcpar-land** [comment]
FWIW, this could potentially be covered by implementing https://github.com/tconbeer/harlequin/discussions/770 and opening the current text buffer in `$EDITOR`.

**Comment by Dking08** [comment]
I have raised a PR for this feature, you can review it here: #977
<img width="1918" height="1078" alt="20260714-2244-04 0791505" src="https://github.com/user-attachments/assets/63961bc2-1566-490b-b75a-ca33b678308c" />
